internal/blockdb/collect.go

```diff
@@ -0,0 +1,70 @@
+package blockdb
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// TxFinder finds transactions given block at height.
+type TxFinder interface {
+	// FindTxs transactions should be in a human-readable format, preferably json.
+	FindTxs(ctx context.Context, height uint64) ([][]byte, error)
+}
+
+// BlockSaver saves transactions for block at height.
+type BlockSaver interface {
+	SaveBlock(ctx context.Context, height int, txs [][]byte) error
+}
+
+// Collector saves block transactions at regular intervals.
+type Collector struct {
+	finder TxFinder
+	log    *zap.Logger
+	rate   time.Duration
+	saver  BlockSaver
+}
+
+// NewCollector creates a valid poller that polls every duration at rate.
+func NewCollector(finder TxFinder, saver BlockSaver, rate time.Duration, log *zap.Logger) *Collector {
+	return &Collector{
+		finder: finder,
+		log:    log,
+		rate:   rate,
+		saver:  saver,
+	}
+}
+
+// Collect saves block transactions starting at height 1 and advancing by 1 height as long as there are
+// no errors with finding or saving the transactions.
+func (p *Collector) Collect(ctx context.Context) {
+	tick := time.NewTicker(p.rate)
+	defer tick.Stop()
+	var height uint64 = 1
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-tick.C:
+			if err := p.saveTxsForHeight(ctx, height); err != nil {
+				p.log.Info("Failed to save transactions", zap.Error(err), zap.Uint64("height", height))
```

Comment:
Might be nice to avoid logging if we can detect that we are trying to access a block that isn't available yet. It might be `ErrHeightExceedsChainHead` or `ErrHeightNotAvailable ` from https://pkg.go.dev/github.com/tendermint/tendermint@v0.35.5/rpc/coretypes#pkg-variables

Reply:
True, but if not a tendermint chain...